### PR TITLE
Adds ability to define a source for atlantis.yaml

### PR DESCRIPTION
**Pull request by mos3abof**
Atlantis currently looks for the global atlantis.yaml configuration file in the branch it pulls down from the PR. This PR adds another flag, `--config-source`, that is passed to the atlantis server, and defines the source of the atlantis.yaml file. This now supports only Github, but can be extended to support other sources in the future,(e.g., Bitbucket, S3, Google Storage, etc.).

To specify a file within a github repository in a given branch you need to add this new flag in the following format:

`--config-source="https://github.com/<owner>/<repo>/[path]atlantis.yaml?[ref=<branch>]"`

Replace <owner>, <repo>, and <branch> with the github username, github repository, and branch name respectively. `[path]` is optional though.

The `--allow-repo-config` parameter is still required for this feature to work. In the future, `--allow-repo-config` can be deprecated and its behaviour can be incorporated in the `--config-source` flag.

### CODE REVIEW

File: server/fetchers/github_fetcher.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+)
+
+type GithubFetcherConfig struct {
+	Repo     string
+	Username string
+	Path     string
+	Token    string
+}
+
+var logger = logging.NewSimpleLogger("server", false, logging.Debug)
```
Comment (mos3abof) [inline server/fetchers/github_fetcher.go]: This needs a better way to be handled.

File: Dockerfile [hunk 2]
```diff
@@ -1,3 +1,9 @@
+# Build Go binary without cgo dependencies
+FROM golang:1.12.1 as builder
+COPY . /go/src/github.com/runatlantis/atlantis
+WORKDIR /go/src/github.com/runatlantis/atlantis
+RUN CGO_ENABLED=0 go build -o atlantis main.go
```
Comment (craigfurman) [inline Dockerfile:5]: Building the binary in the golang image is a legit thing to do, and not just because we needed it for container-builder to work! However, it would be great to separate this into its own commit, so that it can be submitted upstream in a separate PR to the `--config-source` flag

File: Dockerfile [hunk 3]
```diff
@@ -1,3 +1,9 @@
+# Build Go binary without cgo dependencies
+FROM golang:1.12.1 as builder
```
Comment (craigfurman) [inline Dockerfile]:
```suggestion
FROM golang:1.12 as builder
```

autobumping batch versions carries risk, but it's worth it for patch versions IMO

File: Dockerfile [hunk 4]
```diff
@@ -1,3 +1,9 @@
+# Build Go binary without cgo dependencies
```
Comment (craigfurman) [inline Dockerfile]: Why?

File: server/events/project_command_builder.go [hunk 5]
```diff
@@ -2,6 +2,10 @@ package events
 
 import (
 	"fmt"
+	"github.com/runatlantis/atlantis/server/fetchers"
```
Comment (craigfurman) [inline server/events/project_command_builder.go]: Mixing stdlib and non-stdlib imports is unidiomatic in Go. I recommend configuring your editor ro run `goimports` on save and it'll move it for you.

**Review by craigfurman — CHANGES_REQUESTED** [review]
I'd like to see some tests if feasible, and if we want to submit this upstream
the maintainers of Atlantis may want that too.

We try to avoid commit headlines longer than 50 characters. If you haven't
already, check out https://chris.beams.io/posts/git-commit.

IMO, we should attempt to convey intent and context in the following places,
falling back to the next if we can't:

Self-documenting code > Comments > Commit messages > PR messages.

There is a lot of context in the PR message that I think could have been better
placed in the commit message in this case.

About the approach itself, it seems totally reasonable. Atlantis already has
code hosting site specific handling in order to pull code over HTTPS using API
tokens, so it makes sense that we'd need code host specific handling to fetch
the master atlantis.yaml.

Just to stimulate conversation, did we consider a simpler feature, which is
basically a flag `--use-config-from-master`? As far as I know pulling the master
atlantis.yaml is the only use case we can foresee.

Once we get this merged, we should consider submitting PR(s) upstream.

File: .dockerignore [hunk 6]
```diff
@@ -1,3 +1,2 @@
-*
```
Comment (craigfurman) [inline .dockerignore:1]: I reckon we can actually just remove this whole file.

File: Dockerfile (hunk 4 above)
Comment (mos3abof) [reply Dockerfile]: Not sure actually. This bit was copied from the previous PR.

File: .dockerignore (hunk 6 above)
Comment (mos3abof) [reply .dockerignore:1]: Makes sense. Also, separating this into its own commit will make it easier to not include it in the upstream PR.

File: server/events/project_command_builder.go (hunk 5 above)
Comment (mos3abof) [reply server/events/project_command_builder.go]: :+1:

**Comment by mos3abof** [comment]
Checks fails expected, as `Dockerfile` changes commit removed for now. Will add it after finishing the rest of the comments as a separate commit.

**Review by mos3abof — COMMENTED** [review]
In response to Craig's comment:

> I'd like to see some tests if feasible, and if we want to submit this upstream
> the maintainers of Atlantis may want that too.

Added some tests.

> We try to avoid commit headlines longer than 50 characters. If you haven't
> already, check out https://chris.beams.io/posts/git-commit.

Fixed.

> IMO, we should attempt to convey intent and context in the following places,
> falling back to the next if we can't:
> 
> Self-documenting code > Comments > Commit messages > PR messages.
> 
> There is a lot of context in the PR message that I think could have been better
> placed in the commit message in this case.

Makes sense. Done.

> About the approach itself, it seems totally reasonable. Atlantis already has
> code hosting site specific handling in order to pull code over HTTPS using API
> tokens, so it makes sense that we'd need code host specific handling to fetch
> the master atlantis.yaml.
> 
> Just to stimulate conversation, did we consider a simpler feature, which is
> basically a flag `--use-config-from-master`? As far as I know pulling the master
> atlantis.yaml is the only use case we can foresee.

I reckon the effort wouldn't be significantly different, and it would be nicer if other people could end up actually benefitting from this if it goes upstream.


> Once we get this merged, we should consider submitting PR(s) upstream.

That is the plan as far as I know.

File: server/server.go [hunk 7]
```diff
@@ -30,6 +29,9 @@ import (
 	"syscall"
 	"time"
 
+	"github.com/runatlantis/atlantis/server/events/db"
+	"github.com/runatlantis/atlantis/server/fetchers"
```
Comment (benwh) [inline server/server.go]: I'd move these two lines down into the block of imports below (sorted lexicographically), so that they're co-located with the rest of the `runatlantis/atlantis` imports.

File: server/server.go [hunk 8]
```diff
@@ -111,6 +114,28 @@ func NewServer(userConfig UserConfig, config Config) (*Server, error) {
 	var gitlabClient *vcs.GitlabClient
 	var bitbucketCloudClient *bitbucketcloud.Client
 	var bitbucketServerClient *bitbucketserver.Client
+	var fetcherConfig fetchers.FetcherConfig
+
+	if userConfig.ConfigSource != "" {
+		sourceType, err := fetchers.GetType(userConfig.ConfigSource)
+		if err != nil {
+			logger.Err("error determining config source type: %s", err)
```
Comment (benwh) [inline server/server.go]: I don't think that it makes sense to continue with running the server if these steps fail, so we should probably do `return nil, errors.Wrap("...", err)` instead of logging the error? This looks like it would be more consistent with what the rest of the setup code further down in the file does.

Same applies to the next 2 `logger.Err` calls

File: server/server.go [hunk 9]
```diff
@@ -111,6 +114,28 @@ func NewServer(userConfig UserConfig, config Config) (*Server, error) {
...
+		}
+
+		switch sourceType {
+		case fetchers.Github:
+			githubFetcherConfig := fetchers.GithubFetcherConfig{}
+			err := fetchers.ParseConnection(&githubFetcherConfig, userConfig.ConfigSource, userConfig.GithubToken)
```
Comment (benwh) [inline server/server.go]: nit: I find the naming `ParseConnection` a little confusing, as we're not operating on a connection in the traditional sense. `ParseRemoteReference`?
Also applies to the error strings emitted below.

File: server/fetchers/parser.go [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+	"errors"
+	"fmt"
+	"strings"
+)
+
+type ConfigSourceType int
+
+const (
+	Github ConfigSourceType = iota
+	S3
```
Comment (benwh) [inline server/fetchers/parser.go]: I'd suggest removing `S3` and `Bitbucket`, given that they're not implemented. If the intention here is to show that multiple sources can be supported then we can add a comment instead.

File: server/fetchers/parser.go [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+	Bitbucket
+)
+
+type Parser interface {
+	ParseConnection(conn string, token string) error
+}
+
+func GetType(conn string) (ConfigSourceType, error) {
+	s := strings.Split(conn, "://")
```
Comment (benwh) [inline server/fetchers/parser.go]: Could we use `url.Parse` to make this neater?

File: server/fetchers/parser.go [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+
+	if len(s) == 0 {
+		return 0, errors.New("")
+	}
+	switch strings.ToLower(s[0]) {
+	case "github":
```
Comment (benwh) [inline server/fetchers/parser.go]: We're essentially making up a protocol which isn't used anywhere else or [universally recognised](https://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml), which seems a bit off.

What do you think about using https instead, and determining the config provider based on the hostname part of the URI?

File: server/fetchers/github_fetcher.go [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+
+	p := strings.Split(s[1], "/")
+
+	if len(p) < 3 {
+		return errors.New("connection string incomplete")
+	}
+
+	c.Username = p[0]
+	c.Repo = p[1]
+	c.Path = strings.Join(p[2:], "/")
```
Comment (benwh) [inline server/fetchers/github_fetcher.go]: The `Path` field is confusingly named, to me at least, because it will contain not just the path but also the query parameters (i.e. `?ref=master`)

As mentioned in `server/fetchers/parser.go`, parsing as a URL would improve this situation I think, then we could split this out into a separate field which the `formatUrl` func can use.

File: server/fetchers/github_fetcher.go [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+	c.Token = token
+
+	return nil
+}
+
+// Exported for easier testing
+// Returns error, and error code
+func ValidateGithubFetcherConfig(config *GithubFetcherConfig) (error, string) {
```
Comment (benwh) [inline server/fetchers/github_fetcher.go]: I think that the second return parameter gets us nothing here, and can be dropped. I'd also remove the comment on this func!

Comment (benwh) [reply server/fetchers/github_fetcher.go]: Thinking about it, does this function buy us much overall? I reckon it could make more sense to put the validation in `ParseConnection`. If we make that function immediately blow up if it's given incorrect values then that's much better than having started a plan/apply but then having it fail due to incorrect config.

File: server/events/project_command_builder.go [hunk 15]
```diff
@@ -95,6 +101,25 @@ func (p *DefaultProjectCommandBuilder) buildPlanAllCommands(ctx *CommandContext,
 		return nil, err
 	}
 
+	switch p.FetcherConfig.ConfigType {
+	case fetchers.Github:
+		ctx.Log.Debug("fetching github config...")
+		configString, err := p.FetcherConfig.GithubConfig.FetchConfig(ctx.Log)
+		if err != nil {
+			ctx.Log.Err("error fetching github config: %s", err)
```
Comment (benwh) [inline server/events/project_command_builder.go]: Rather than logging an error we should wrap and return it immediately, to be more consistent with the rest of the code in this file.

This is pretty important in this case - if we fail to retrieve the config from the remote source then we should abort; going down a different code path is confusing and potentially dangerous depending on the atlantis config. e.g. The server may allow applying unapproved pull requests by default, but the repository that we're operating on should not.

 (Also applies to the error logged below).

File: server/events/project_command_builder.go [hunk 16]
```diff
@@ -95,6 +101,25 @@ func (p *DefaultProjectCommandBuilder) buildPlanAllCommands(ctx *CommandContext,
...
+		}
+
+		// If the request returned content, save it to file, otherwise delete the existing file.
+		configFilePath := filepath.Join(repoDir, yaml.AtlantisYAMLFilename)
+		if configString != "" {
+			err = ioutil.WriteFile(configFilePath, []byte(configString), 0644)
+		} else {
+			err = os.Remove(configFilePath)
```
Comment (benwh) [inline server/events/project_command_builder.go]: Following on from the comment above: this code path seems like a bit of a trap, unless there's a good reason for it that I'm missing? Are we better off by just aborting?

File: server/fetchers/github_fetcher.go [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+
+	if config.Token == "" {
+		return errors.New("github token is not set"), invalidGithubToken
+	}
+
+	return nil, validObject
+}
+
+// Exported for easier testing.
+func FormatUrl(config *GithubFetcherConfig) string {
```
Comment (benwh) [inline server/fetchers/github_fetcher.go]: We don't need to export these functions, because we're testing them in the same package :)

File: server/fetchers/github_fetcher.go [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+		logSink.Err("invalid github fetcher config: %s", err)
+		return
+	}
+
+	// If the config is valid, construct the url to get the config from
+	url := FormatUrl(c)
+	logSink.Debug("Getting config from: %s", url)
+
+	// Perform the actual request
+	body, err := DoRequest(url, c.Token)
```
Comment (benwh) [inline server/fetchers/github_fetcher.go]: The result of err isn't being checked here, we need to do something with it

File: server/fetchers/github_fetcher.go [hunk 19]
```diff
@@ -0,0 +1,184 @@
...
+	invalidGithubRepoName string = "3 invalid github repo name"
+	invalidGithubPath     string = "4 invalid github path"
+	invalidGithubToken    string = "5 invalid github token"
+)
+
+func (c *GithubFetcherConfig) FetchConfig(logSink *logging.SimpleLogger) (configString string, err error) {
+	// Validate the fetcher config before moving forward
+	err, code := ValidateGithubFetcherConfig(c)
+	if err != nil || code != validObject {
+		logSink.Err("invalid github fetcher config: %s", err)
```
Comment (benwh) [inline server/fetchers/github_fetcher.go]: This should just return a wrapped error, rather than logging and then returning the un-wrapped error. Same applies throughout this func.

File: server/fetchers/github_fetcher.go [hunk 20]
```diff
@@ -0,0 +1,184 @@
...
+	return ioutil.ReadAll(resp.Body)
+}
+
+// Exported for easier testing.
+func ParseResponse(body []byte, dat *map[string]interface{}) error {
+	var err error
+	if err = json.Unmarshal(body, &dat); err != nil {
+		return err
+	}
+	_, err = json.MarshalIndent(dat, "", " ")
```
Comment (benwh) [inline server/fetchers/github_fetcher.go]: What's this for?

File: server/fetchers/github_fetcher.go [hunk 21]
```diff
@@ -0,0 +1,184 @@
...
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	return ioutil.ReadAll(resp.Body)
+}
+
+// Exported for easier testing.
+func ParseResponse(body []byte, dat *map[string]interface{}) error {
```
Comment (benwh) [inline server/fetchers/github_fetcher.go]: This function is just wrapping `json.Unmarshal`. What do you think about merging `ValidateParsedMap` (and maybe also `DecodeContent`) into this func?

It would remove a bit of error handling logic in the `FetchConfig` function.

File: server/fetchers/github_fetcher.go [hunk 22]
```diff
@@ -0,0 +1,184 @@
...
+	return nil, validObject
+}
+
+// Exported for easier testing.
+func FormatUrl(config *GithubFetcherConfig) string {
+	return fmt.Sprintf(GithubAPIURLFormat, config.Username, config.Repo, config.Path)
+}
+
+// Exported for easier testing.
+func DoRequest(url string, token string) ([]byte, error) {
```
Comment (benwh) [inline server/fetchers/github_fetcher.go]: The repo already imports the `go-github` package. Can we use this instead? I think it'd cut out a lot of the manual error handling and validation checks. It'd probably also make testing easier! https://godoc.org/github.com/google/go-github/github#RepositoriesService.GetContents

File: server/fetchers/github_fetcher.go [hunk 23]
```diff
@@ -0,0 +1,184 @@
...
+func ValidateParsedMap(parsedMap map[string]interface{}) error {
+	typ, ok := parsedMap["type"]
+	if !ok {
+		return errors.New("syntax error in response, no 'type' provided")
+	}
+
+	// @TODO: Support symlink from Github
+	// If not a file, skip.
+	if typ != "file" {
+		return nil
```
Comment (benwh) [inline server/fetchers/github_fetcher.go]: Returning nil, rather than an error, seems like a bad idea to me.

This becomes irrelevant if we use go-github instead though.

File: server/fetchers/github_fetcher_test.go [hunk 24]
```diff
@@ -0,0 +1,160 @@
+package fetchers
+
+import (
+	"testing"
+)
+
+func TestParseConnection(t *testing.T) {
```
Comment (benwh) [inline server/fetchers/github_fetcher_test.go]: I haven't gone over these tests in too much detail, as I think some of the changes proposed in the non-test files will render some of the tests redundant.

However the one pattern that stands out to me, which should be corrected, is that the helpers in `testing/assertions.go` should be used to be consistent with the rest of the codebase.

File: server/fetchers/parser_test.go [hunk 25]
```diff
@@ -0,0 +1,55 @@
+package fetchers
+
+import "testing"
+
+func TestParserNormalScenario(t *testing.T) {
```
Comment (benwh) [inline server/fetchers/parser_test.go]: There's quite a bit of duplication between these tests; can we switch them to be [table-driven](https://github.com/golang/go/wiki/TableDrivenTests) instead? There's a few other tests in the codebase which use that pattern.

**Review by benwh — COMMENTED** [review]
Overall this looks like it's taking the right shape, nice work so far!

There's quite a few comments I'm leaving, but most of are for fairly minor things and hopefully preempting the feedback we'll receive when trying to push this upstream later on.

On top of those, there's a few errors which are picked up by the linter. `golangci-lint run` will point these out to you!

The only other thing I'm slightly worried about is whether there's an impedance mismatch between what we've got here and the direction that Atlantis is taking: This code effectively supports one set of Atlantis configuration across _all_ repositories that you're running Atlantis on. However the [server-side config](https://www.runatlantis.io/docs/server-side-repo-config.html) change that was merged in a couple of weeks back allows for definitions of different configuration per repository, with wildcard matches.
So *ideally* we would add a key into the server-side configuration, within the object that goes into the `repos` array, that defines the remote config source - allowing one to specify different sources per repo.

However I don't think that we should worry about that at this stage (consider the previous paragraph a brain-dump, before I lose context on this) - I propose we focus on getting this PR merged using the current approach, and then we can tackle the wider problem if we need to when we try to upstream this work.

File: server/events/project_command_builder.go (hunk 15 above)
Comment (mos3abof) [reply server/events/project_command_builder.go]: Agreed. I was a bit torn apart between both approaches.

File: server/events/project_command_builder.go (hunk 16 above)
Comment (mos3abof) [reply server/events/project_command_builder.go]: Makes sense.

File: server/fetchers/github_fetcher.go (hunk 22 above)
Comment (mos3abof) [reply server/fetchers/github_fetcher.go]: The problem is Github wouldn't allow `git-archive`, which would've been ideal to get just the file we need. So the alternative is cloning the whole repo, check out the correct banch/commit, and get the file from there. This seemed like an overkill.

Comment (mos3abof) [reply server/fetchers/github_fetcher.go]: Scratch that reply. Your comment makes sense now.

**Review by benwh — APPROVED** [review]
Looking good now, let's give this a run!
